docs/content/Schema/Dynamic-Schema-Creation.md

```diff
@@ -0,0 +1,268 @@
+category: Data Schema
+menuOrder: 2
+---
+
+Cube.js allows schemas to be created on-the-fly using a special
+[`asyncModule()`][ref-async-module] function only available in the [schema
+execution environment][ref-schema-env]. `asyncModule()` allows registering an
+async function to be executed at the end of the data schema compile phase so
+additional definitions can be added. This is often useful in situations where
+schema properties can be dynamically updated through an API, for example.
+
+<!-- prettier-ignore-start -->
+[[warning | Note]]
+| Each `asyncModule` call will be invoked only once per schema compilation.
+<!-- prettier-ignore-end -->
+
+[ref-schema-env]: /schema-execution-environment
+[ref-async-module]: /schema-execution-environment#asyncmodule
+
+When creating schemas via `asyncModule()`, it is important to be aware of the
+following differences compared to statically defining schemas with `cube()`:
+
+- The `sql` and `drillMembers` properties for both dimensions and measures must
+  be of type `() => string`
+
+Cube.js supports importing JavaScript logic from other files in a schema, so it
+is useful to declare utility functions for handling the above differences in a
+separate file:
+
+[ref-import-export]: /export-import
+
+```javascript
+// schema/utils.js
+export const convertStringPropToFunction = (propNames, dimensionDefinition) => {
+  let newResult = { ...dimensionDefinition };
+  propNames.forEach((propName) => {
+    const propValue = newResult[propName];
+
+    if (!propValue) {
+      return;
+    }
+
+    newResult[propName] = () => propValue;
+  });
+  return newResult;
+};
+
+export const transformDimensions = (dimensions) => {
+  return Object.keys(dimensions).reduce((result, dimensionName) => {
+    const dimensionDefinition = dimensions[dimensionName];
+    return {
+      ...result,
+      [dimensionName]: convertStringPropToFunction(
+        ['sql'],
+        dimensionDefinition
+      ),
+    };
+  }, {});
+};
+
+export const transformMeasures = (measures) => {
+  return Object.keys(measures).reduce((result, dimensionName) => {
+    const dimensionDefinition = measures[dimensionName];
+    return {
+      ...result,
+      [dimensionName]: convertStringPropToFunction(
+        ['sql, drillMembers'],
+        dimensionDefinition
+      ),
+    };
+  }, {});
+};
+```
+
+## Generation
+
+In the following example, we retrieve a JSON object representing all our cubes
+using `fetch()`, transform some of the properties to be functions that return a
+string, and then finally use the [`cube()` global function][ref-globals] to
+generate schemas from that data:
+
+[ref-globals]: /schema-execution-environment#cube-js-globals-cube-and-others
+
+```javascript
+// schema/DynamicSchema.js
+const fetch = require('node-fetch');
+import {
+  convertStringPropToFunction,
+  transformDimensions,
+  transformMeasures,
+} from './utils';
+
+asyncModule(async () => {
+  const dynamicCubes = await (
```

Comment:
@hassankhan Should we write somewhere pretty simple notice that JSON doesn't support serialization for functions?

Reply:
@ovr I tried to explain that on lines 23-27, perhaps I could be clearer?